src/psyclone/gocean1p0.py

```diff
@@ -1906,89 +1906,86 @@ def gen_ocl(self, indent=0):
-            if len(symbol.shape) > 0:
-                code = code + "__global "
-
-            if symbol.datatype == "real":
-                code = code + "double "
-            elif symbol.datatype == "integer":
-                code = code + "int "
-            elif symbol.datatype == "character":
-                code = code + "char "
-            else:
-                raise NotImplementedError(
-                    "Type '{0}' OpenCL declaration not supported."
-                    "".format(symbol.datatype))
 
+            # If argument is an array, it is allocated to the OpenCL global
+            # address space.
             if len(symbol.shape) > 0:
-                code = code + "* restrict "
-                array_arguments.append((symbol.name, len(symbol.shape)))
+                code = code + "__global "
 
-            code = code + symbol.name + ",\n"
+            code = code + symbol.gen_c_definition() + ",\n"
 
         code = code[:-2]   # Remove last ",\n"
         code = code + "\n" + self.indent(indent + 1) + "){\n"
 
         # Declare local variables.
-        for symbol in self.symbol_table._symbols.values():
-            if symbol not in self.symbol_table.argument_list:
-                code = code + self.indent(indent + 1)
-                if symbol.datatype == 'real':
-                    code = code + "double "
-                elif symbol.datatype == 'integer':
-                    code = code + "int "
-                elif symbol.datatype == 'character':
-                    code = code + "char "
-                else:
-                    raise NotImplementedError(
-                        "Type '{0}' OpenCL declaration not supported."
-                        "".format(symbol.datatype))
-
-                if len(symbol.shape) > 0:
-                    code = code + "* "
-                code = code + symbol.name + ";\n"
-
-        # Generate a LEN variable for each array.
-        for name, dimensions in array_arguments:
-            for dim in range(2, dimensions + 1):
-                code = code + self.indent(indent + 1) + "int " + name
-                code = code + "LEN" + str(dim) + " = get_global_size("
+        for symbol in self.symbol_table.local_symbols:
+            code = code + self.indent(indent + 1)
+            code = code + symbol.gen_c_definition() + ";\n"
+
+        # Declare a <name>LEN<DIM> variable for each array dimension.
+        # In OpenCL the sizes are retrived from the kernel global_work_size
+        # (e.g. "int arrayLEN1 = get_global_size(1);")
+        for symbol in self.symbol_table.argument_list[2:]:
+            dimensions = len(symbol.shape)
+            for dim in range(1, dimensions + 1):
+                code = code + self.indent(indent + 1) + "int "
+                varname = symbol.name + "LEN" + str(dim)
+
+                # Check there is no clash with other variables
+                if varname in self.symbol_table:
+                    raise GenerationError(
```

Comment:
Please could you add a TODO here and reference your new Issue on SymbolTable/NameSpaceManager inter-comparison.

Reply:
Same. I added the TODO in the generic class.